src/Compilers/Core/Portable/Operations/OperationExtensions.cs

```diff
@@ -322,5 +322,84 @@ internal static IOperation GetRootOperation(this IOperation operation)
+        /// Applicable kinds: <see cref="BranchKind.Break"/> and <see cref="BranchKind.Continue"/>.</exception>
+        public static ILoopOperation GetCorrespondingLoop(this IBranchOperation branchOperation)
+        {
+            if (branchOperation == null)
+            {
+                throw new ArgumentNullException(nameof(branchOperation));
+            }
+
+            if (branchOperation.SemanticModel == null)
+            {
+                throw new InvalidOperationException(CodeAnalysisResources.OperationMustNotBeControlFlowGraphPart);
+            }
+
+            if (branchOperation.BranchKind != BranchKind.Break && branchOperation.BranchKind != BranchKind.Continue)
+            {
+                throw new InvalidOperationException(
+                    string.Format(CodeAnalysisResources.InvalidBranchKindForFindingCorrespondingLoop, branchOperation.BranchKind));
+            }
+
+            return FindCorrespondingOperation<ILoopOperation>(branchOperation, 
+                op => op is ISwitchOperation && branchOperation.BranchKind == BranchKind.Break);
+        }
+
+        /// <summary>
+        /// Gets a switch operation that corresponds to the given branch operation.
+        /// </summary>
+        /// <param name="branchOperation">The branch operation for which a corresponding switch is looked up</param>
+        /// <returns>The corresponding switch operation or <c>null</c> in case not found (e.g. no switch syntax or the branch
+        /// belongs to loop instead of switch operation)</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="branchOperation"/> is null</exception>
+        /// <exception cref="InvalidOperationException">The operation is a part of Control Flow Graph or it has an invalid branch kind. 
+        /// Applicable kinds: <see cref="BranchKind.Break"/>.</exception>
+        public static ISwitchOperation GetCorrespondingSwitch(this IBranchOperation branchOperation)
+        {
+            if (branchOperation == null)
+            {
+                throw new ArgumentNullException(nameof(branchOperation));
+            }
+
+            if (branchOperation.SemanticModel == null)
+            {
+                throw new InvalidOperationException(CodeAnalysisResources.OperationMustNotBeControlFlowGraphPart);
+            }
+
+            if (branchOperation.BranchKind != BranchKind.Break)
+            {
+                throw new InvalidOperationException(
+                    string.Format(CodeAnalysisResources.InvalidBranchKindForFindingCorrespondingSwitch, branchOperation.BranchKind));
+            }
+
+            return FindCorrespondingOperation<ISwitchOperation>(branchOperation, op => op is ILoopOperation);
+        }
+
+        private static T FindCorrespondingOperation<T>(IOperation operation, Func<IOperation, bool> shouldStopAscending) where T : IOperation
```

Comment:
Would the implementation be simpler if you walk up the parent chain and directly compare `IBranchOperation.Target` with `ILoopOperation.ExitLabel` or  `ILoopOperation.ContinueLabel` or `ISwitchOperation.ExitLabel`? At the minimum you should add an assert that for a non-null return value of `GetCorrespondingSwitch` and `GetCorrespondingLoop`, the label symbols match.